### PR TITLE
shebangs: use `/usr/bin/env bash` instead of `/bin/bash` for BSD compat

### PR DESCRIPTION
**Pull request by gasche**
Currently our CI machine on FreeBSD fails to build with the following
error message:

```
bytecomp/generate_runtimedef.sh runtime/caml/fail.h runtime/primitives > bytecomp/runtimedef.ml
/bin/sh: bytecomp/generate_runtimedef.sh: not found
```

I'm not sure what the problem is, but one guess is that this file uses a `#!/bin/bash` shebang which appears [not to work](https://forums.freebsd.org/threads/cannot-run-bash-script-file.12600/) in some FreeBSD environments.

This PR converts all the files with a `/bin/bash` shebang to
`/usr/bin/env bash`. Some files use `/bin/sh` instead, and they are
unchanged. It may be the case that some 'bash' files could in fact run
using 'sh', but checking this requires more work and I decided to
follow the original author's intent.

(Personally I like using bash explicitly as it gives a consistent
programming environment that I can easily test on my machine; but this
should not be at the expense of portability.)

### CODE REVIEW

**Comment by shindere** [comment]
Hi Gabriel!

Many thanks for having taken the time to look into this.

Re: using `/bin/sh` everywhere, you say it would be more work to make
sure it would work. Would there be something else to do except trying
these scripts with dash?

If such a test is enough to make sure they are Posix, wouldn't it be
worthi it? What do you and others think?

**Comment by gasche** [comment]
Neither `dash` nor `bash --posix` actually guarantee that they will reject non-POSIX programs, so they do not allow me to *test* that scripts are portable. (Besides, testing programs is difficult as they may contain hard-to-reach error-handling code, etc.) On the other hand, I don't know of a single compatibility problem in the distribution caused by differing `bash` version, so I'm confident that using `bash` gives me a consistent programming environment.

(If we have to spend time on shell scripts, I would rather play with looking at the output of `shellcheck`, fixing the alarms, and considering some of the lints/warnings.)

**Comment by gasche** [comment]
P.S.: We had a similar discussion in the May/June 2017 caml-devel thread "Questions on make's shell selection", and my understanding is that there was a moderate consensus that using `bash` was a reasonable thing to do. (David was especially in favor.)

File: bytecomp/generate_runtimedef.sh [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/env bash
```
Comment (dra27) [inline bytecomp/generate_runtimedef.sh]: This can be changed to `/bin/sh`

File: runtime/gen_primitives.sh [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/usr/bin/env bash
```
Comment (dra27) [inline runtime/gen_primitives.sh]: This can be changed to `/bin/sh` (although it's needed for Dune side, so it's less important to change)

**Review by shindere — APPROVED** [review]

**Comment by shindere** [comment]
Just approved. You may want to add a Change entry.

I was wondering why the script in the testsuite didn't fail. Any idea?

**Review by dra27 — APPROVED** [review]
I've suggested two scripts which are clearly `sh` because they are in the build pipeline. For the rest I think it's fine to have the bash requirement.

Switching to `sh` in `bytecomp/generate_runtimedef.sh` means that bash is no longer a build requirement of OCaml, it's just a requirement for two utility scripts in tools and various developer tools (including the testsuite).

**Comment by dra27** [comment]
@shindere - I assume the BSD worker doesn't have `afl-showmap`, so the testsuite script won't have been executed.

**Comment by dra27** [comment]
Up to you on Changes - I didn't feel it was necessary!

**Comment by dra27** [comment]
That said, having a Changes entry for it would allow BSD users to bash us for any regressions in future, either on shebangs or for requiring bash unnecessarily in the build!

**Comment by gasche** [comment]
Well, we don't want to commit to not requiring `bash` for the build, given that we almost decided to set `SHELL=bash` in our makefile last year (I think we didn't purely out of inertia). I'll still downgrade the two scripts you pointed.

**Comment by shindere** [comment]
David Allsopp (2018/10/16 01:27 -0700):
> @shindere - I assume the BSD worker doesn't have `afl-showmap`, so the
> testsuite script won't have been executed.

Sure, makes ssense of course, thanks and sorry for having looked at it
in a too shallow way.

**Comment by dra27** [comment]
> Well, we don't want to commit to not requiring `bash` for the build, given that we almost decided to set `SHELL=bash` in our makefile last year (I think we didn't purely out of inertia). I'll still downgrade the two scripts you pointed.

I agree, I was being the proverbial devil's advocate :wink:

**Comment by gasche** [comment]
(It looks like merging this PR did indeed fix the build on BSD machines.)

**Comment by shindere** [comment]
Gabriel Scherer (2018/10/16 08:07 -0700):
> (It looks like merging this PR did indeed fix the build on BSD
> machines.)

Yes indeed! Thanks alot again for having taken the time to investigate
this!

**Pull request merged**